### ISSUE TITLE
Parser runs into endless loop

### ISSUE THREAD

**Issue by nikku** [lezer-parser/lezer#9]
I've [created a lezer-based parser for FEEL](https://github.com/nikku/feelin/blob/master/src/feel.grammar), a domain specific language. My purpose is to experiment with lezers error recovery features to try it out and provide editor assistance for the language. 

__Bug Description__

The language allows parentheses `(a)` as well as intervals `(a..b)` starting with a `(` character. When parsing the incomplete input `(a` the parser runs in an endless loop and eventually exausts the host memory.

At this point I'm out of ideas how to approach and track down this issue. Do you have any pointers?

__Parse Log__

The loop can be seen via the parse log produced via `LOG=parse ...`:

```
[0] -> [0,7] (via shift for "(" @ 0)
[0,7] -> [0,7,3] (via shift for Name @ 1)
[0,7,3] -> [0,7,3,37] (via reduce of ("." Name)*-wrap for ␄ @ 2)
[0,7,3,37] -> [0,7,42] (via always-reduce QualifiedName)
[0,7,42] -> [0,7,43] (via forced reduction at eof)
[0,7,43] -> [0,7,40] (via forced reduction at eof)
[0,7,40] -> [0,7,11] (via forced reduction at eof)
[0,7,11] -> [0,7,43] (via always-reduce textualExpression)
[0,7,43] -> [0,7,40] (via forced reduction at eof)
[0,7,40] -> [0,7,11] (via forced reduction at eof)
[0,7,11] -> [0,7,43] (via always-reduce textualExpression)
[0,7,43] -> [0,7,40] (via forced reduction at eof)
[0,7,40] -> [0,7,11] (via forced reduction at eof)
[0,7,11] -> [0,7,43] (via always-reduce textualExpression)
[0,7,43] -> [0,7,40] (via forced reduction at eof)
[0,7,40] -> [0,7,11] (via forced reduction at eof)
[0,7,11] -> [0,7,43] (via always-reduce textualExpression)
[0,7,43] -> [0,7,40] (via forced reduction at eof)
[0,7,40] -> [0,7,11] (via forced reduction at eof)
...
```

__Expected Behavior__

If it is due to a problem / ambiguity in my grammar, I'd expect the generator to warn me about the grammar problem.

__How to Reproduce__

```sh
git clone git@github.com:nikku/feelin.git -b infinite-loop-error-case
cd feelin
npm install 
npm run build-parser
npm test
```

**Comment by marijnh** [comment]
This appears to be a case of the error-recovery strategy failing to make progress. I've reduced it to the following grammar

```
@precedence { call }

@top[name=Script] { a+ }

a { b | "{" "x" "}" }

b { a !call "{" "}" | "x" | "(" b ")" }
```

The precedence for `!call`, along with the mutual recursion between `a` and `b` (which are `expression` and `textualExpression` in your grammar) creates a state machine that throws the error recovery for a loop. I'm still investigating why exactly it goes wrong like that and how it should be addressed.

**Comment by marijnh** [comment]
Attached patch, released as 0.5.1, should help.

**Comment by nikku** [comment]
Thank you. This fixes the issue.

Great work!
